quota/quota.go

```diff
@@ -0,0 +1,180 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package quota
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+)
+
+// Group is a quota group of snaps, services or sub-groups that are all subject
+// to specific resource quotas. Current quota resource types supported are just
+// memory, but this can be expanded in the future.
+type Group struct {
+	// Name is the name of the quota group. Certain names are reserved and have
+	// special meaning such as user.* or system.* or snapd.*, but are otherwise
+	// generically choosable by a user. This name will correspond to the name of
+	// the systemd slice underlying the quota group.
+	Name string
+
+	// SubGroups is the set of sub-groups that are subject to this quota.
+	// Sub-groups have their own limits, subject to the requirement that the
+	// highest quota for a sub-group is that of the parent group.
+	SubGroups []*Group `json:"sub-groups,omitempty"`
+
+	// MemoryLimit is the limit of memory available to the processes in the
+	// group where if the total used memory of all the processes exceeds the
+	// limit, oom-killer is invoked which will start killing processes. The
+	// specific behavior of which processes are killed is subject to the
+	// ExhaustionBehavior.
+	MemoryLimit quantity.Size `json:"memory-limit,omitempty"`
```

Comment:
systemd modern name for this is MemoryMax. I wonder given though that we have separate behavior, whether we should use a name that is not used by systemd? otoh we it depends also what name/combination we will use in the API/command line.

Reply:
I have no strong opinions on the name we use here right now